## 43% of end-end tests have passed
```
Total executed tests: 7
Failed tests: 3
Expected to fail tests: 1
Passed tests: 3
```
```
IDs of failed tests: 727230,703133,727229 
```
```
IDs of expected to fail tests: 727232 
```
<h3>Failed tests (3)</h3><details><summary>Click to expand</summary><br/><li><a href="https://ci.status.im/job/status-mobile/job/e2e/job/status-app-prs-rerun/parambuild/?BRANCH_NAME=bugfix/20358&PR_ID=20711&APK_NAME=20711.apk&TR_CASE_IDS=727230,703133,727229">Rerun failed tests</a></li><br/><h4>Class TestCommunityOneDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3619343">test_restore_multiaccount_with_waku_backup_remove_switch</a>, id: 703133 </b></td></tr><tr><td><p><blockquote><div>## Signed in successfully!</div><div>Device 1: Find `CommunitiesTab` by `accessibility id`: `communities-stack-tab` </div></blockquote></p>

```
critical/chats/test_public_chat_browsing.py:222: in test_restore_multiaccount_with_waku_backup_remove_switch
    self.home.communities_tab.click()
../views/base_element.py:90: in click
    element = self.find_element()
../views/base_element.py:79: in find_element
    raise NoSuchElementException(
 Device 1: CommunitiesTab by accessibility id: `communities-stack-tab` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/951d49c2495c48e8a23dd56ccb21ce5f?auth=c2d73dc763cb05d3e884c6effb4ce795#75">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestWalletMultipleDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3619345">test_wallet_send_asset_from_drawer</a>, id: 727230 </b></td></tr><tr><td><p><blockquote><div>Device 2: Tap on found: Button </div><div>Device 2: Find `ChatsTab` by `accessibility id`: `chats-stack-tab` </div></blockquote></p>

```
critical/test_wallet.py:155: in test_wallet_send_asset_from_drawer
    self.wallet_2.chats_tab.click()
../views/base_element.py:90: in click
    element = self.find_element()
../views/base_element.py:79: in find_element
    raise NoSuchElementException(
 Device 2: ChatsTab by accessibility id: `chats-stack-tab` is not found on the screen; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#no-such-element-exception
```

<br/><br/></td></tr><tr><td><b>2. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3619344">test_wallet_send_eth</a>, id: 727229 </b></td></tr><tr><td><p><blockquote><div>Device 1: Find `Button` by `accessibility id`: `top-bar` </div><div>Device 2: Find `Button` by `accessibility id`: `top-bar` </div></blockquote></p>

```
critical/test_wallet.py:148: in test_wallet_send_eth
    self.errors.verify_no_errors()
base_test_case.py:191: in verify_no_errors
    pytest.fail('\n '.join([self.errors.pop(0) for _ in range(len(self.errors))]))
 Sender balance is not updated on Etherscan, it is 0.0 but expected to be -0.0001
E    Receiver balance is not updated on Etherscan, it is 0.0026 but expected to be 0.0027
```

<br/><br/></td></tr></tbody></table></details><h3>Expected to fail tests (1)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestWalletOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3619347">test_wallet_add_remove_watch_only_account</a>, id: 727232 </b></td></tr><tr><td><p><blockquote><div>Device 1: Find `EditBox` by `accessibility id`: `add-address-to-watch` </div><div>Device 1: Type `0x8d2413447ff297d30bdc475f6d5cb00254685aae` to `EditBox` </div></blockquote></p>

```
critical/test_wallet.py:239: in test_wallet_add_remove_watch_only_account
    self.wallet_view.add_watch_only_account(address=address_to_watch, account_name=new_account_name)
../views/wallet_view.py:150: in add_watch_only_account
    self.account_has_activity_label.wait_for_visibility_of_element()
../views/base_element.py:147: in wait_for_visibility_of_element
    raise TimeoutException(
 Device 1: Text by accessibility id:`account-has-activity` is not found on the screen after wait_for_visibility_of_element 
```

<b>[[Missing networks in account address, https://github.com/status-im/status-mobile/issues/20166]]</b><br/><br/><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/f14318fbf99d47559ed3ba7f343b2e11?auth=354c80f23b244b6c82af07621f5a3ee7#96">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details><h3>Passed tests (3)</h3><details><summary>Click to expand</summary><br/><br/><h4>Class TestCommunityOneDeviceMerged:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3619341">test_community_copy_and_paste_message_in_chat_input</a>, id: 702742 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/951d49c2495c48e8a23dd56ccb21ce5f?auth=c2d73dc763cb05d3e884c6effb4ce795#46">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestOneToOneChatMultipleSharedDevicesNewUi:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3619342">test_1_1_chat_non_latin_messages_stack_update_profile_photo</a>, id: 702745 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/56382dfa9b13412a9ddeb747817ea243?auth=69424e134a1977e5f5af91e9801aa8fc#89">Steps, video, logs</a></li></ul></p><p>Device 2:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/f4b83569755345ada62955a26e122a0b?auth=30cf1c7e36c8043e932de044e3f45aef#78">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table><h4>Class TestWalletOneDevice:</h4><table style="width: 100%"><colgroup><col span="1" style="width: 20%;"><col span="1" style="width: 80%;"></colgroup><tbody><tr></tr><tr><td><b>1. <a href="https://ethstatus.testrail.net/index.php?/tests/view/3619346">test_wallet_add_remove_regular_account</a>, id: 727231 </b></td></tr><tr><td><ins>Device sessions</ins><p><ul><p>Device 1:<ul><li><a href="https://eu-central-1.saucelabs.com/jobs/f14318fbf99d47559ed3ba7f343b2e11?auth=354c80f23b244b6c82af07621f5a3ee7#32">Steps, video, logs</a></li></ul></p></ul></p></td></tr></tbody></table></details>

--- a/src/quo/components/drawers/bottom_actions/style.cljs
+++ b/src/quo/components/drawers/bottom_actions/style.cljs
@@ -21,12 +21,14 @@
                        (colors/theme-colors colors/white colors/neutral-95 theme))})
 
 (defn buttons-container
-  [actions]
-  {:flex-direction     (if (= actions :two-vertical-actions) :column :row)
-   :justify-content    :space-around
-   :padding-vertical   12
-   :gap                12
-   :padding-horizontal 20})
+  [actions container-style]
+  (merge
+   {:flex-direction     (if (= actions :two-vertical-actions) :column :row)
+    :justify-content    :space-around
+    :padding-vertical   12
+    :gap                12
+    :padding-horizontal 20}
+   container-style))
 
 (def description-top
   {:flex-direction  :row

--- a/src/quo/components/drawers/bottom_actions/view.cljs
+++ b/src/quo/components/drawers/bottom_actions/view.cljs
@@ -30,7 +30,8 @@
       [:button-two-props {:optional true} [:maybe :map]]
       [:scroll? {:optional true} [:maybe :boolean]]
       [:blur? {:optional true} [:maybe :boolean]]
-      [:container-style {:optional true} [:maybe :map]]]]]
+      [:container-style {:optional true} [:maybe :map]]
+      [:buttons-container-style {:optional true} [:maybe :map]]]]]
    :any])
 
 (def ^:private role-icon
@@ -42,7 +43,7 @@
 (defn- view-internal
   [{:keys [actions description description-text description-top-text error-message role button-one-label
            button-two-label blur? button-one-props button-two-props scroll? container-style
-           context-tag-props]}]
+           buttons-container-style context-tag-props]}]
   (let [theme (quo.theme/use-theme)]
     [rn/view
      {:style (merge (style/container scroll? blur? theme) container-style)}
@@ -72,7 +73,7 @@
             :context (i18n/label (keyword "t" role))}
            context-tag-props)]])
 
-     [rn/view {:style (style/buttons-container actions)}
+     [rn/view {:style (style/buttons-container actions buttons-container-style)}
       (when (or (= actions :two-actions)
                 (= actions :two-vertical-actions))
         [button/button

--- a/src/status_im/constants.cljs
+++ b/src/status_im/constants.cljs
@@ -287,6 +287,7 @@
 (def ^:const wallet-connect-supported-events #{"accountsChanged" "chainChanged"})
 (def ^:const wallet-connect-session-proposal-event "session_proposal")
 (def ^:const wallet-connect-session-request-event "session_request")
+(def ^:const wallet-connect-session-delete-event "session_delete")
 (def ^:const wallet-connect-user-rejected-error-key "USER_REJECTED")
 
 (def ^:const transaction-pending-type-wallet-connect-transfer "WalletConnectTransfer")

--- a/src/status_im/contexts/shell/qr_reader/view.cljs
+++ b/src/status_im/contexts/shell/qr_reader/view.cljs
@@ -117,13 +117,14 @@
 (defn- f-internal-view
   []
   (let [{:keys [keyboard-shown]} (hooks/use-keyboard)]
-    [:<>
-     (when keyboard-shown
-       (rn/dismiss-keyboard!))
-     [scan-qr-code/view
-      {:title           (i18n/label :t/scan-qr)
-       :share-button?   true
-       :on-success-scan on-qr-code-scanned}]]))
+    (rn/use-mount
+     (fn []
+       (when keyboard-shown
+         (rn/dismiss-keyboard!))))
+    [scan-qr-code/view
+     {:title           (i18n/label :t/scan-qr)
+      :share-button?   true
+      :on-success-scan on-qr-code-scanned}]))
 
 (defn view
   []

--- a/src/status_im/contexts/wallet/connected_dapps/view.cljs
+++ b/src/status_im/contexts/wallet/connected_dapps/view.cljs
@@ -13,26 +13,26 @@
     [utils.re-frame :as rf]))
 
 (defn- on-disconnect
-  [wallet-account {:keys [name topic]}]
+  [wallet-account {:keys [name topic pairing-topic]}]
   (rf/dispatch [:hide-bottom-sheet])
   (rf/dispatch
    [:wallet-connect/disconnect-dapp
-    {:topic      topic
-     :on-success (fn []
-                   (rf/dispatch [:wallet-connect/remove-pairing-by-topic topic])
-                   (rf/dispatch [:toasts/upsert
-                                 {:id   :dapp-disconnect-success
-                                  :type :positive
-                                  :text (i18n/label :t/disconnect-dapp-success
-                                                    {:dapp    name
-                                                     :account (:name wallet-account)})}]))
-     :on-fail    (fn []
-                   (rf/dispatch [:toasts/upsert
-                                 {:id   :dapp-disconnect-failure
-                                  :type :negative
-                                  :text (i18n/label :t/disconnect-dapp-fail
-                                                    {:dapp    name
-                                                     :account (:name wallet-account)})}]))}]))
+    {:topic         topic
+     :pairing-topic pairing-topic
+     :on-success    (fn []
+                      (rf/dispatch [:toasts/upsert
+                                    {:id   :dapp-disconnect-success
+                                     :type :positive
+                                     :text (i18n/label :t/disconnect-dapp-success
+                                                       {:dapp    name
+                                                        :account (:name wallet-account)})}]))
+     :on-fail       (fn []
+                      (rf/dispatch [:toasts/upsert
+                                    {:id   :dapp-disconnect-failure
+                                     :type :negative
+                                     :text (i18n/label :t/disconnect-dapp-fail
+                                                       {:dapp    name
+                                                        :account (:name wallet-account)})}]))}]))
 
 (defn- on-dapp-disconnect-press
   [wallet-account dapp]
@@ -83,7 +83,8 @@
   []
   (let [{:keys [bottom]}                   (safe-area/get-insets)
         {:keys [color] :as wallet-account} (rf/sub [:wallet/current-viewing-account])
-        pairings                           (rf/sub [:wallet-connect/pairings])
+        sessions                           (rf/sub
+                                            [:wallet-connect/sessions-for-current-account])
         theme                              (quo.theme/use-theme)]
     [rn/view {:flex 1}
      [header
@@ -91,7 +92,7 @@
        :wallet-account wallet-account
        :on-close       #(rf/dispatch [:navigate-back])
        :on-add         #(rf/dispatch [:navigate-to :screen/wallet.scan-dapp])}]
-     (if (empty? pairings)
+     (if (empty? sessions)
        [quo/empty-state
         {:title           (i18n/label :t/no-dapps)
          :description     (i18n/label :t/no-dapps-description)
@@ -99,16 +100,16 @@
          :container-style style/empty-container-style}]
        [rn/view (style/dapps-container bottom)
         [rn/flat-list
-         {:data                    pairings
+         {:data                    sessions
           :always-bounce-vertical  false
           :content-container-style (style/dapps-list theme)
-          :render-fn               (fn [{:keys                    [topic]
-                                         {:keys [icons name url]} :peerMetadata}]
+          :render-fn               (fn [{:keys [topic pairingTopic name url iconUrl]}]
                                      [quo/dapp
-                                      {:dapp                {:avatar (get icons 0)
-                                                             :name   name
-                                                             :value  url
-                                                             :topic  topic}
+                                      {:dapp                {:avatar        iconUrl
+                                                             :name          name
+                                                             :value         url
+                                                             :topic         topic
+                                                             :pairing-topic pairingTopic}
                                        :accessibility-label (str "dapp-" topic)
                                        :state               :default
                                        :action              :icon

--- a/src/status_im/contexts/wallet/wallet_connect/core.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/core.cljs
@@ -85,12 +85,34 @@
 
 (defn required-networks-supported?
   [proposal supported-networks]
-  (let [required-networks (get-in proposal [:params :requiredNamespaces :eip155 :chains])
-        supported-eip155  (set (map chain-id->eip155 supported-networks))]
-    (every? #(contains? supported-eip155 %) required-networks)))
+  (let [supported-namespaces #{:eip155}
+        required-namespaces  (get-in proposal [:params :requiredNamespaces])]
+    (when (every? #(contains? supported-namespaces %)
+                  (keys required-namespaces))
+      (let [required-networks (get-in required-namespaces [:eip155 :chains])
+            supported-eip155  (set (map chain-id->eip155 supported-networks))]
+        (every? #(contains? supported-eip155 %)
+                required-networks)))))
 
 (defn get-networks-by-mode
   [db]
   (let [test-mode? (get-in db [:profile/profile :test-networks-enabled?])
         networks   (get-in db [:wallet :networks (if test-mode? :test :prod)])]
     (mapv #(-> % :chain-id) networks)))
+
+(defn event-should-be-handled?
+  [db {:keys [topic]}]
+  (some #(= topic %)
+        (map :topic (:wallet-connect/sessions db))))
+
+(defn sdk-session->db-session
+  [{:keys [topic expiry pairingTopic] :as session}]
+  {:topic        topic
+   :expiry       expiry
+   :sessionJson  (transforms/clj->json session)
+   :pairingTopic pairingTopic
+   :name         (get-in session [:peer :metadata :name])
+   :iconUrl      (get-in session [:peer :metadata :icons 0])
+   :url          (get-in session [:peer :metadata :url])
+   :accounts     (get-in session [:namespaces :eip155 :accounts])
+   :disconnected false})

--- a/src/status_im/contexts/wallet/wallet_connect/effects.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/effects.cljs
@@ -32,13 +32,6 @@
      :handler     handler})))
 
 (rf/reg-fx
- :effects.wallet-connect/fetch-pairings
- (fn [{:keys [web3-wallet on-success on-fail]}]
-   (-> (wallet-connect/get-pairings web3-wallet)
-       (promesa/then on-success)
-       (promesa/catch on-fail))))
-
-(rf/reg-fx
  :effects.wallet-connect/pair
  (fn [{:keys [web3-wallet url on-success on-fail]}]
    (when web3-wallet
@@ -50,13 +43,6 @@
  :effects.wallet-connect/disconnect
  (fn [{:keys [web3-wallet topic on-success on-fail]}]
    (-> (wallet-connect/core-pairing-disconnnect web3-wallet topic)
-       (promesa/then on-success)
-       (promesa/catch on-fail))))
-
-(rf/reg-fx
- :effects.wallet-connect/fetch-active-sessions
- (fn [{:keys [web3-wallet on-success on-fail]}]
-   (-> (wallet-connect/get-active-sessions web3-wallet)
        (promesa/then on-success)
        (promesa/catch on-fail))))
 
@@ -73,6 +59,13 @@
            :approved-namespaces approved-namespaces})
          (promesa/then on-success)
          (promesa/catch on-fail)))))
+
+(rf/reg-fx
+ :effects.wallet-connect/fetch-active-sessions
+ (fn [{:keys [web3-wallet on-success on-fail]}]
+   (-> (wallet-connect/get-active-sessions web3-wallet)
+       (promesa/then on-success)
+       (promesa/catch on-fail))))
 
 (rf/reg-fx
  :effects.wallet-connect/sign-message

--- a/src/status_im/contexts/wallet/wallet_connect/events.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/events.cljs
@@ -8,7 +8,8 @@
             status-im.contexts.wallet.wallet-connect.responding-events
             [status-im.contexts.wallet.wallet-connect.utils :as wc-utils]
             [taoensso.timbre :as log]
-            [utils.i18n :as i18n]))
+            [utils.i18n :as i18n]
+            [utils.transforms :as types]))
 
 (rf/reg-event-fx
  :wallet-connect/init
@@ -22,12 +23,7 @@
  (fn [{:keys [db]} [web3-wallet]]
    {:db (assoc db :wallet-connect/web3-wallet web3-wallet)
     :fx [[:dispatch [:wallet-connect/register-event-listeners]]
-         [:effects.wallet-connect/fetch-pairings
-          {:web3-wallet web3-wallet
-           :on-fail     #(log/error "Failed to get dApp pairings" {:error %})
-           :on-success  (fn [data]
-                          (rf/dispatch [:wallet-connect/set-pairings
-                                        (js->clj data :keywordize-keys true)]))}]]}))
+         [:dispatch [:wallet-connect/fetch-persisted-sessions]]]}))
 
 (rf/reg-event-fx
  :wallet-connect/register-event-listeners
@@ -40,7 +36,11 @@
            [:effects.wallet-connect/register-event-listener
             [web3-wallet
              constants/wallet-connect-session-request-event
-             #(rf/dispatch [:wallet-connect/on-session-request %])]]]})))
+             #(rf/dispatch [:wallet-connect/on-session-request %])]]
+           [:effects.wallet-connect/register-event-listener
+            [web3-wallet
+             constants/wallet-connect-session-delete-event
+             #(rf/dispatch [:wallet-connect/on-session-delete %])]]]})))
 
 (rf/reg-event-fx
  :wallet-connect/on-init-fail
@@ -54,7 +54,10 @@
  (fn [{:keys [db]} [proposal]]
    (log/info "Received Wallet Connect session proposal: " {:id (:id proposal)})
    (let [accounts                     (get-in db [:wallet :accounts])
-         without-watched              (remove :watch-only? (vals accounts))
+         current-viewing-address      (get-in db [:wallet :current-viewing-account-address])
+         available-accounts           (filter #(and (:operable? %)
+                                                    (not (:watch-only? %)))
+                                              (vals accounts))
          networks                     (wallet-connect-core/get-networks-by-mode db)
          session-networks             (wallet-connect-core/proposal-networks-intersection proposal
                                                                                           networks)
@@ -65,9 +68,10 @@
                     :wallet-connect/current-proposal assoc
                     :request                         proposal
                     :session-networks                session-networks
-                    :address                         (-> without-watched
-                                                         first
-                                                         :address))
+                    :address                         (or current-viewing-address
+                                                         (-> available-accounts
+                                                             first
+                                                             :address)))
         :fx [[:dispatch
               [:open-modal :screen/wallet.wallet-connect-session-proposal]]]}
        {:fx [[:dispatch
@@ -85,9 +89,16 @@
 
 (rf/reg-event-fx
  :wallet-connect/on-session-request
- (fn [_ [event]]
-   (log/info "Received Wallet Connect session request: " event)
-   {:fx [[:dispatch [:wallet-connect/process-session-request event]]]}))
+ (fn [{:keys [db]} [event]]
+   (when (wallet-connect-core/event-should-be-handled? db event)
+     {:fx [[:dispatch [:wallet-connect/process-session-request event]]]})))
+
+(rf/reg-event-fx
+ :wallet-connect/on-session-delete
+ (fn [{:keys [db]} [{:keys [topic] :as event}]]
+   (when (wallet-connect-core/event-should-be-handled? db event)
+     (log/info "Received Wallet Connect session delete: " event)
+     {:fx [[:dispatch [:wallet-connect/disconnect-session topic]]]})))
 
 (rf/reg-event-fx
  :wallet-connect/reset-current-session-proposal
@@ -105,27 +116,17 @@
    {:db (dissoc db :wallet-connect/current-request)}))
 
 (rf/reg-event-fx
- :wallet-connect/set-pairings
- (fn [{:keys [db]} [pairings]]
-   {:db (assoc db :wallet-connect/pairings pairings)}))
-
-(rf/reg-event-fx
- :wallet-connect/remove-pairing-by-topic
- (fn [{:keys [db]} [topic]]
-   {:db (update db
-                :wallet-connect/pairings
-                (fn [pairings]
-                  (remove #(= (:topic %) topic) pairings)))}))
-
-(rf/reg-event-fx
  :wallet-connect/disconnect-dapp
- (fn [{:keys [db]} [{:keys [topic on-success on-fail]}]]
+ (fn [{:keys [db]} [{:keys [pairing-topic on-success on-fail]}]]
    (let [web3-wallet (get db :wallet-connect/web3-wallet)]
      {:fx [[:effects.wallet-connect/disconnect
             {:web3-wallet web3-wallet
-             :topic       topic
+             :topic       pairing-topic
              :on-fail     on-fail
-             :on-success  on-success}]]})))
+             :on-success  (fn []
+                            (rf/dispatch [:wallet-connect/disconnect-session pairing-topic])
+                            (when on-success
+                              (on-success)))}]]})))
 
 (rf/reg-event-fx
  :wallet-connect/pair
@@ -136,15 +137,6 @@
              :url         url
              :on-fail     #(log/error "Failed to pair with dApp" {:error %})
              :on-success  #(log/info "dApp paired successfully")}]]})))
-
-(rf/reg-event-fx
- :wallet-connect/fetch-active-sessions
- (fn [{:keys [db]}]
-   (let [web3-wallet (get db :wallet-connect/web3-wallet)]
-     {:fx [[:effects.wallet-connect/fetch-active-sessions
-            {:web3-wallet web3-wallet
-             :on-fail     #(log/error "Failed to get active sessions" {:error %})
-             :on-success  #(log/info "Got active sessions successfully" {:sessions %})}]]})))
 
 (rf/reg-event-fx
  :wallet-connect/approve-session
@@ -204,10 +196,55 @@
                                          {:version version}))}]]]}
        {:fx [[:dispatch [:wallet-connect/pair scanned-text]]]}))))
 
+;; We first load sessions from database, then we initiate a call to Wallet Connect SDK and
+;; then replace the list we have stored in the database with the one that came from the SDK.
+;; In addition to that, we also update the backend state by marking sessions that are not
+;; active anymore by calling `:wallet-connect/disconnect-session`.
+(rf/reg-event-fx
+ :wallet-connect/fetch-active-sessions-success
+ (fn [{:keys [db now]} [sessions]]
+   (let [persisted-sessions (:wallet-connect/sessions db)
+         sessions           (->> (js->clj sessions :keywordize-keys true)
+                                 vals
+                                 (map wallet-connect-core/sdk-session->db-session))
+         expired-sessions   (remove
+                             (fn [{:keys [expiry]}]
+                               (> expiry (/ now 1000)))
+                             persisted-sessions)]
+     {:fx (mapv (fn [{:keys [pairingTopic]}]
+                  [:wallet-connect/disconnect-session pairingTopic])
+                expired-sessions)
+      :db (assoc db :wallet-connect/sessions sessions)})))
+
+(rf/reg-event-fx
+ :wallet-connect/fetch-active-sessions
+ (fn [{:keys [db]}]
+   (let [web3-wallet (get db :wallet-connect/web3-wallet)]
+     {:fx [[:effects.wallet-connect/fetch-active-sessions
+            {:web3-wallet web3-wallet
+             :on-fail     #(log/error "Failed to get active sessions" {:error %})
+             :on-success  #(rf/dispatch [:wallet-connect/fetch-active-sessions-success %])}]]})))
+
 (rf/reg-event-fx
  :wallet-connect/fetch-persisted-sessions-success
  (fn [{:keys [db]} [sessions]]
-   {:db (assoc db :wallet-connect/persisted-sessions sessions)}))
+   (let [sessions' (mapv (fn [{:keys [sessionJson] :as session}]
+                           (assoc session
+                                  :accounts
+                                  (-> sessionJson
+                                      types/json->clj
+                                      :namespaces
+                                      :eip155
+                                      :accounts)))
+                         sessions)]
+     {:fx [[:dispatch [:wallet-connect/fetch-active-sessions]]]
+      :db (assoc db :wallet-connect/sessions sessions')})))
+
+(rf/reg-event-fx
+ :wallet-connect/fetch-persisted-sessions-fail
+ (fn [_ [error]]
+   (log/info "Wallet Connect fetch persisted sessions failed" error)
+   {:fx [[:dispatch [:wallet-connect/fetch-active-sessions]]]}))
 
 (rf/reg-event-fx
  :wallet-connect/fetch-persisted-sessions
@@ -218,7 +255,7 @@
             ;; 0 means, return everything
             :params     [0]
             :on-success [:wallet-connect/fetch-persisted-sessions-success]
-            :on-error   #(log/info "Wallet Connect fetch persisted sessions failed" %)}]]]}))
+            :on-error   [:wallet-connect/fetch-persisted-sessions-fail]}]]]}))
 
 (rf/reg-event-fx
  :wallet-connect/persist-session
@@ -226,5 +263,22 @@
    {:fx [[:json-rpc/call
           [{:method     "wallet_addWalletConnectSession"
             :params     [(js/JSON.stringify session-info)]
-            :on-success #(log/info "Wallet Connect session persisted")
+            :on-success (fn []
+                          (log/info "Wallet Connect session persisted")
+                          (rf/dispatch [:wallet-connect/fetch-persisted-sessions]))
+            :on-error   #(log/info "Wallet Connect session persistence failed" %)}]]]}))
+
+(rf/reg-event-fx
+ :wallet-connect/disconnect-session
+ (fn [{:keys [db]} [pairing-topic]]
+   {:db (update db
+                :wallet-connect/sessions
+                (fn [sessions]
+                  (->> sessions
+                       (remove #(= (:pairingTopic %) pairing-topic))
+                       (into []))))
+    :fx [[:json-rpc/call
+          [{:method     "wallet_disconnectWalletConnectSession"
+            :params     [pairing-topic]
+            :on-success #(log/info "Wallet Connect session disconnected")
             :on-error   #(log/info "Wallet Connect session persistence failed" %)}]]]}))

--- a/src/status_im/contexts/wallet/wallet_connect/modals/common/header/view.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/modals/common/header/view.cljs
@@ -10,12 +10,12 @@
    [quo/text
     {:size   :heading-1
      :weight :semi-bold}
-    (let [{:keys [name icons]} (:peerMetadata dapp)]
+    (let [{:keys [name iconUrl]} dapp]
       [rn/view {:style style/header-dapp-name}
        [quo/summary-tag
         {:type         :dapp
          :label        name
-         :image-source (first icons)}]])
+         :image-source iconUrl}]])
     (str " " label " ")
     (let [{:keys [emoji customization-color name]} account]
       [rn/view {:style style/header-account-name}

--- a/src/status_im/contexts/wallet/wallet_connect/responding_events.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/responding_events.cljs
@@ -132,7 +132,7 @@
              :proposal    current-proposal
              :on-success  #(log/info "Wallet Connect session proposal rejected")
              :on-error    #(log/error "Wallet Connect unable to reject session proposal")}]
-           [:dispatch [:wallet-connect/reset-current-session]]]})))
+           [:dispatch [:wallet-connect/reset-current-session-proposal]]]})))
 
 ;; NOTE: Currently we only reject a session if the user rejected it
 ;; But this needs to be solidified to ensure other cases:

--- a/src/status_im/contexts/wallet/wallet_connect/session_proposal/style.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/session_proposal/style.cljs
@@ -6,12 +6,13 @@
    :padding-top        12})
 
 (def approval-note-container
-  {:margin-horizontal 20
-   :padding           12
-   :border-radius     16
-   :border-width      1
-   :border-color      colors/neutral-10
-   :background-color  colors/neutral-2_5})
+  {:margin-horizontal  20
+   :padding-horizontal 16
+   :padding-vertical   12
+   :border-radius      16
+   :border-width       1
+   :border-color       colors/neutral-10
+   :background-color   colors/neutral-2_5})
 
 (def approval-note-title
   {:color         colors/neutral-50
@@ -20,10 +21,8 @@
 (def approval-note-li
   {:flex           1
    :flex-direction :row
-   :align-items    :center})
-
-(def approval-li-spacer
-  {:width 8})
+   :align-items    :center
+   :gap            8})
 
 (def account-switcher-title
   {:padding-horizontal 20})
@@ -31,3 +30,6 @@
 (def account-switcher-list
   {:margin-top         8
    :padding-horizontal 8})
+
+(def footer-buttons-container
+  {:padding-horizontal 0})

--- a/src/status_im/contexts/wallet/wallet_connect/session_proposal/view.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/session_proposal/view.cljs
@@ -33,16 +33,21 @@
         labels    [(i18n/label :t/check-your-account-balance-and-activity)
                    (i18n/label :t/request-txns-and-message-signing)]]
     [rn/view {:style style/approval-note-container}
-     [quo/text {:style style/approval-note-title}
+     [quo/text
+      {:style  style/approval-note-title
+       :weight :regular
+       :size   :paragraph-2}
       (i18n/label :t/dapp-will-be-able-to {:dapp-name dapp-name})]
      (map-indexed
       (fn [idx label]
         ^{:key (str idx label)}
         [rn/view {:style style/approval-note-li}
          [quo/icon :i/bullet
-          {:color colors/neutral-50}]
-         [rn/view {:style style/approval-li-spacer}]
-         [quo/text label]])
+          {:color colors/neutral-40}]
+         [quo/text
+          {:weight :regular
+           :size   :paragraph-2}
+          label]])
       labels)]))
 
 (defn- format-network-name
@@ -57,16 +62,14 @@
 
 (defn- accounts-list
   []
-  (let [accounts         (rf/sub [:wallet/accounts-without-watched-accounts])
+  (let [accounts         (rf/sub [:wallet/operable-accounts-without-watched-accounts])
         selected-address (rf/sub [:wallet-connect/current-proposal-address])]
     [rn/view {:style style/account-switcher-list}
-     (for [account accounts]
-       ^{:key (-> account :address str)}
+     (for [{:keys [address] :as account} accounts]
+       ^{:key (str address)}
        [quo/account-item
         {:type          :default
-         :state         (if (and selected-address
-                                 (= (account :address)
-                                    selected-address))
+         :state         (if (= address selected-address)
                           :selected
                           :default)
          :account-props account
@@ -101,12 +104,10 @@
                                                       (map format-network-name)
                                                       (string/join ", "))
         network-images                           (mapv :source session-networks)
-        data-item-common-props                   {:blur?       false
-                                                  :description :default
-                                                  :card?       false
-                                                  :label       :preview
-                                                  :status      :default
-                                                  :size        :large}
+        data-item-common-props                   {:blur?  false
+                                                  :card?  false
+                                                  :status :default
+                                                  :size   :large}
         account-data-item-props                  (assoc data-item-common-props
                                                         :right-content {:type :accounts
                                                                         :size :size-32
@@ -116,9 +117,7 @@
                                                         :on-press      show-account-switcher-bottom-sheet
                                                         :title         (i18n/label :t/account-title)
                                                         :subtitle      name
-                                                        :icon-right?   true
-                                                        :right-icon    :i/chevron-right
-                                                        :icon-color    colors/neutral-10)
+                                                        :right-icon    :i/chevron-right)
         networks-data-item-props                 (assoc data-item-common-props
                                                         :right-content {:type :network
                                                                         :data network-images}
@@ -136,18 +135,21 @@
   []
   (let [customization-color (rf/sub [:profile/customization-color])]
     [quo/bottom-actions
-     {:actions          :two-actions
-      :button-two-label (i18n/label :t/decline)
-      :button-two-props {:type                :grey
-                         :accessibility-label :wc-deny-connection
-                         :on-press            #(do (rf/dispatch [:navigate-back])
-                                                   (rf/dispatch
-                                                    [:wallet-connect/reject-session-proposal]))}
-      :button-one-label (i18n/label :t/connect)
-      :button-one-props {:customization-color customization-color
-                         :type                :primary
-                         :accessibility-label :wc-connect
-                         :on-press            #(rf/dispatch [:wallet-connect/approve-session])}}]))
+     {:actions                 :two-actions
+      :buttons-container-style style/footer-buttons-container
+      :button-two-label        (i18n/label :t/decline)
+      :button-two-props        {:type                :grey
+                                :accessibility-label :wc-deny-connection
+                                :on-press            (fn []
+                                                       (rf/dispatch [:navigate-back])
+                                                       (rf/dispatch
+                                                        [:wallet-connect/reject-session-proposal]))}
+      :button-one-label        (i18n/label :t/connect)
+      :button-one-props        {:customization-color customization-color
+                                :type                :primary
+                                :accessibility-label :wc-connect
+                                :on-press            #(rf/dispatch
+                                                       [:wallet-connect/approve-session])}}]))
 
 (defn- header
   []

--- a/src/status_im/contexts/wallet/wallet_connect/utils.cljs
+++ b/src/status_im/contexts/wallet/wallet_connect/utils.cljs
@@ -11,14 +11,14 @@
 
 (defn timestamp-expired?
   [expiry-timestamp]
-  (> (current-timestamp) expiry-timestamp))
+  (when expiry-timestamp
+    (> (current-timestamp) expiry-timestamp)))
 
 (defn valid-wc-uri?
   [parsed-uri]
-  (let [{:keys [topic version expiryTimestamp]} parsed-uri]
+  (let [{:keys [topic version]} parsed-uri]
     (and (seq topic)
-         (number? version)
-         (number? expiryTimestamp))))
+         (number? version))))
 
 (defn valid-uri?
   "Check if the uri is in the wallet-connect format.

--- a/src/status_im/subs/root.cljs
+++ b/src/status_im/subs/root.cljs
@@ -169,7 +169,7 @@
 (reg-root-key-sub :wallet-connect/web3-wallet :wallet-connect/web3-wallet)
 (reg-root-key-sub :wallet-connect/current-proposal :wallet-connect/current-proposal)
 (reg-root-key-sub :wallet-connect/current-request :wallet-connect/current-request)
-(reg-root-key-sub :wallet-connect/pairings :wallet-connect/pairings)
+(reg-root-key-sub :wallet-connect/sessions :wallet-connect/sessions)
 
 ;;biometrics
 (reg-root-key-sub :biometrics :biometrics)

--- a/src/status_im/subs/wallet/wallet_connect.cljs
+++ b/src/status_im/subs/wallet/wallet_connect.cljs
@@ -1,5 +1,6 @@
 (ns status-im.subs.wallet.wallet-connect
-  (:require [re-frame.core :as rf]
+  (:require [clojure.string :as string]
+            [re-frame.core :as rf]
             [status-im.contexts.wallet.common.utils :as wallet-utils]
             [status-im.contexts.wallet.common.utils.networks :as networks]
             [status-im.contexts.wallet.wallet-connect.core :as wallet-connect-core]
@@ -37,13 +38,23 @@
 (rf/reg-sub
  :wallet-connect/current-request-dapp
  :<- [:wallet-connect/current-request]
- :<- [:wallet-connect/pairings]
- (fn [[request pairings]]
+ :<- [:wallet-connect/sessions]
+ (fn [[request sessions]]
    (let [dapp-url (get-in request [:event :verifyContext :verified :origin])]
-     (->> pairings
-          (filter (fn [pairing]
-                    (= dapp-url (get-in pairing [:peerMetadata :url]))))
+     (->> sessions
+          (filter (fn [session]
+                    (= dapp-url (get session :url))))
           (first)))))
+
+(rf/reg-sub
+ :wallet-connect/sessions-for-current-account
+ :<- [:wallet-connect/sessions]
+ :<- [:wallet/current-viewing-account-address]
+ (fn [[sessions address]]
+   (filter
+    (fn [{:keys [accounts]}]
+      (some #(string/includes? % address) accounts))
+    sessions)))
 
 (rf/reg-sub
  :wallet-connect/current-request-network

